server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -412,11 +420,109 @@ public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task,
                     return;
                 }
             }
-            // fork the execution in the search thread pool
-            runAsync(getExecutor(shard), () -> executeQueryPhase(orig, task), l);
+            ensureAfterSeqNoRefreshed(shard, orig, () -> executeQueryPhase(orig, task), l);
```

Comment:
I think we need the refresh to happen before the lines above. Notice that `rewriteAndFetchShardRequest` does the search idle refresh if necessary and that we then check if no docs match and early terminate. We risk that early termination to result in no results while waiting for a refresh would have given us results.

Alternatively, the change in `canMatch` should always assume a match when there are `waitForCheckpoints`.